javascript/ql/src/semmle/javascript/dataflow/Nodes.qll

```diff
@@ -219,17 +219,38 @@ class GlobalVarRefNode extends DataFlow::ValueNode, DataFlow::DefaultSourceNode
   string getName() { result = astNode.getName() }
 }
 
+/**
+ * Gets a data flow node corresponding to an access to the global object, including
+ * `this` expressions outside functions, references to global variables `window`
+ * and `global`, and uses of the `global` npm package.
+ */
+DataFlow::SourceNode globalObjectRef() {
+  // top-level `this`
+  exists (ThisNode globalThis | result = globalThis |
+    not exists(globalThis.getBinder())
+  )
+  or
+  // DOM
+  result = globalVarRef("window") or
+  // Node.js
+  result = globalVarRef("global") or
+  // `require("global")`
+  result = moduleImport("global")
+}
+
 /**
  * Gets a data flow node corresponding to an access to global variable `name`,
- * either directly or through `window` or `global`.
+ * either directly, through `window` or `global`, or through the `global` npm package.
  */
 pragma[nomagic]
 DataFlow::SourceNode globalVarRef(string name) {
-  result.(GlobalVarRefNode).getName() = name or
-  // DOM environment
-  result = globalVarRef("window").getAPropertyReference(name) or
-  // Node.js environment
-  result = globalVarRef("global").getAPropertyReference(name)
+  result.(GlobalVarRefNode).getName() = name
+  or
+  result = globalObjectRef().getAPropertyReference(name)
+  or
+  // `require("global/document")` or `require("global/window")`
+  (name = "document" or name = "window") and
+  result = moduleImport("global/" + name)
```

Comment:
What about `moduleMember("global", name)` for ES6-imports?

Reply:
I don't think that works; it's implemented as a group of CommonJS modules.

Reply:
I see, I thought maybe you could do `import { window } from "global";`

Reply:
I can't see how that would work. Without transpiling you'd get an error (since ES2015 modules cannot import CommonJS ones), and with transpiling it most likely wouldn't work either since `window` is the default export of [window.js](https://github.com/Raynos/global/blob/master/window.js) (the package's main module), not a named export.